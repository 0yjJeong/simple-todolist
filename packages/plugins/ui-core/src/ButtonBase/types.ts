import React from 'react';

export type ButtonBaseType = {
  palette: {
    /**
     * A font color
     */
    color: React.CSSProperties['color'];
    /**
     * A theme color
     */
    light: React.CSSProperties['background'];
    /**
     * A theme color when it is hovered
     */
    darker: React.CSSProperties['background'];
  };
  /**
   * Text sizes
   */
  size?: 'small' | 'large';
  /**
   * Types of button variation
   */
  variant?: 'text' | 'outlined' | 'contained';
};
