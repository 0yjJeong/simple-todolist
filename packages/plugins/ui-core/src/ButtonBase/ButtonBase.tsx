import styled, { css } from 'styled-components';
import { ButtonBaseType } from './types';

const ButtonBase = styled.button<ButtonBaseType>`
  border: none;
  background: none;
  cursor: pointer;
  padding: 12px 24px;
  border-radius: 5px;
  font-size: ${(props) => (props.size === 'large' ? '1.2' : '0.8')}rem;

  ${(props) => {
    if (props.variant === 'text') {
      return css`
        color: ${props.palette.color};
        :hover {
          background: ${props.palette.darker};
        }
      `;
    }
    if (props.variant === 'contained') {
      return css`
        color: ${props.palette.color};
        background: ${props.palette.light};
        :hover {
          background: ${props.palette.darker};
        }
      `;
    }
    if (props.variant === 'outlined') {
      return css`
        color: ${props.palette.color};
        border: ${`1px solid ${props.palette.light}`};
        :hover {
          background: ${props.palette.darker};
        }
      `;
    }
  }}
`;

export default ButtonBase;
