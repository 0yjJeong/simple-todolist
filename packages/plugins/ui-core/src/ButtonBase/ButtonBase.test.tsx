import { render } from '@testing-library/react';
import React from 'react';
import ButtonBase from './ButtonBase';
import 'jest-styled-components';

const palette = { color: '#ffffff', light: '#2ac955', darker: '#0abf3a' };

describe('<ButtonBase/>', () => {
  it('should render a text button', () => {
    const rendered = render(
      <ButtonBase variant='text' size='small' palette={palette} />
    );
    expect(rendered.container.firstChild).toMatchSnapshot();
    expect(rendered.container.firstChild).toHaveStyleRule(
      'color',
      palette.color
    );
    expect(rendered.container.firstChild).not.toHaveStyleRule(
      'background',
      palette.light
    );
    expect(rendered.container.firstChild).toHaveStyleRule(
      'background',
      palette.darker,
      {
        modifier: '&:hover',
      }
    );
  });

  it('should render an contained button', () => {
    const rendered = render(
      <ButtonBase variant='contained' size='small' palette={palette} />
    );
    expect(rendered.container.firstChild).toMatchSnapshot();
    expect(rendered.container.firstChild).toHaveStyleRule(
      'color',
      palette.color
    );
    expect(rendered.container.firstChild).toHaveStyleRule(
      'background',
      palette.light
    );
    expect(rendered.container.firstChild).toHaveStyleRule(
      'background',
      palette.darker,
      {
        modifier: '&:hover',
      }
    );
  });

  it('should render an outlined button', () => {
    const rendered = render(
      <ButtonBase variant='outlined' size='small' palette={palette} />
    );
    expect(rendered.container.firstChild).toMatchSnapshot();
    expect(rendered.container.firstChild).toHaveStyleRule(
      'color',
      palette.color
    );
    expect(rendered.container.firstChild).toHaveStyleRule(
      'border',
      `1px solid ${palette.light}`
    );
    expect(rendered.container.firstChild).toHaveStyleRule(
      'background',
      palette.darker,
      {
        modifier: '&:hover',
      }
    );
  });
});
