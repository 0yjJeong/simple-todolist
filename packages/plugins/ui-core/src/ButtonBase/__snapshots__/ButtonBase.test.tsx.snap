// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonBase/> should render a text button 1`] = `
.c0 {
  border: none;
  background: none;
  cursor: pointer;
  padding: 12px 24px;
  border-radius: 5px;
  font-size: 0.8rem;
  color: #ffffff;
}

.c0:hover {
  background: #0abf3a;
}

<button
  class="c0"
/>
`;

exports[`<ButtonBase/> should render an contained button 1`] = `
.c0 {
  border: none;
  background: none;
  cursor: pointer;
  padding: 12px 24px;
  border-radius: 5px;
  font-size: 0.8rem;
  color: #ffffff;
  background: #2ac955;
}

.c0:hover {
  background: #0abf3a;
}

<button
  class="c0"
/>
`;

exports[`<ButtonBase/> should render an outlined button 1`] = `
.c0 {
  border: none;
  background: none;
  cursor: pointer;
  padding: 12px 24px;
  border-radius: 5px;
  font-size: 0.8rem;
  color: #ffffff;
  border: 1px solid #2ac955;
}

.c0:hover {
  background: #0abf3a;
}

<button
  class="c0"
/>
`;
